File: leetcodesln/RotateArray.cs
namespace leetcodesln
{
    public class RotateArray
    {
        public void Rotate(int[] nums, int k)
        {
            k %= nums.Length;

            var arr2 = new int[nums.Length * 2];

            for (int i = 0; i < arr2.Length; i++)
            {
                if (i < nums.Length) arr2[i] = nums[i];

                else arr2[i] = nums[i - nums.Length]; ;
            }
            for (int i = 0; i < nums.Length; i++) nums[i] = arr2[nums.Length - k + i];
        }
    }
}

File: Go/questions/RotateArray.go
package questions

// method 1 make a new array and store the rotated result, T O(n) S O(n)
func Rotate(nums []int, k int) {
	n := len(nums)
	res := make([]int, n)

	for i, v := range nums {
		res[(i + k) % n] = v
	}
	copy(nums, res)
}

// method2 reverse array in place, T O(n) S O(1), optimal solution
func RotateInPlace(nums []int, k int) {
	n := k % len(nums)
	reverse(nums, 0, len(nums)-1)
	reverse(nums,0, n-1)
	reverse(nums, n, len(nums)-1)
}

func reverse(nums []int, l int, r int) {
	for l < r {
		nums[l], nums[r] = nums[r], nums[l]
		l++
		r--
	}
}
